File: src/com/designpatterns/creational/prototype/ClassRoom.java
package com.designpatterns.creational.prototype;

import java.util.ArrayList;
import java.util.List;

public class ClassRoom implements IClassRoom {

    private final IStudentDAO studentDAO;
    private List<Student> students = new ArrayList<>();

    ClassRoom(IStudentDAO studentDAO) {
        this.studentDAO = studentDAO;
    }

    @Override
    public void loadClassRoom() {
        students = studentDAO.getStudents();
    }

    @Override
    public void addStudent(Student student) {
        students.add(student);
    }

    @Override
    public List<Student> getStudents() {
        return students;
    }

    @Override
    public IClassRoom clone() {
        IClassRoom cloneClassRoom = new ClassRoom(studentDAO);
        for (Student student : students) {
            cloneClassRoom.addStudent(student);
        }
        return cloneClassRoom;
    }
}

File: src/com/designpatterns/creational/prototype/kotlin/ClassRoom.kt
package com.designpatterns.creational.prototype.kotlin

import java.util.ArrayList

class ClassRoom internal constructor(private val studentDAO: IStudentDAO) : IClassRoom {

    override var students: MutableList<Student> = ArrayList()

    override fun loadClassRoom() {
        students = studentDAO.getStudentsList() as MutableList<Student>
    }

    override fun addStudent(student: Student) {
        students.add(student)
    }

    override fun clone(): IClassRoom {
        val cloneClassRoom: IClassRoom = ClassRoom(studentDAO)
        for (student in students) {
            cloneClassRoom.addStudent(student)
        }
        return cloneClassRoom
    }
}